File: framework/languages/java/samples/java/DeliveryDispatch/Server/Tracking/src/main/java/systems/zlink/samples/deliverydispatch/server/tracking/spots/DeliveryTrackingSpot.java
package systems.zlink.samples.deliverydispatch.server.tracking.spots;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.IOException;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import systems.zlink.contracts.messaging.Message;
import systems.zlink.framework.CancellationToken;
import systems.zlink.framework.actors.ZLinkActor;
import systems.zlink.framework.spots.ZLinkSpot;
import systems.zlink.framework.spots.ZLinkSpotActorJoinResponse;
import systems.zlink.framework.spots.ZLinkSpotContext;
import systems.zlink.framework.spots.ZLinkSpotCreateResponse;
import systems.zlink.samples.deliverydispatch.shared.contracts.Messages;

public final class DeliveryTrackingSpot implements ZLinkSpot {
    private final ZLinkSpotContext context;
    private final DeliverySpotDirectory directory;
    private final ObjectMapper json;
    private final Map<String, ZLinkActor> customers = new LinkedHashMap<>();
    private final List<Messages.DeliveryStatusChanged> history = new ArrayList<>();
    private String deliveryId = "";

    public DeliveryTrackingSpot(
        ZLinkSpotContext context,
        DeliverySpotDirectory directory,
        ObjectMapper json) {
        this.context = context;
        this.directory = directory;
        this.json = json;
    }

    @Override
    public ZLinkSpotContext context() {
        return context;
    }

    @Override
    public ZLinkSpotCreateResponse onCreate(Message request) {
        Messages.DeliverySpotCreate create = decode(request, Messages.DeliverySpotCreate.class);
        this.deliveryId = create.deliveryId();
        directory.add(deliveryId, this);
        return ZLinkSpotCreateResponse.accept(encode(new Messages.DeliverySpotCreated(deliveryId)));
    }

    @Override
    public ZLinkSpotActorJoinResponse onActorJoin(
        ZLinkActor actor,
        Message request,
        CancellationToken cancellationToken) {
        Messages.DeliverySpotJoin join = decode(request, Messages.DeliverySpotJoin.class);
        if (!join.deliveryId().equals(deliveryId)) {
            return ZLinkSpotActorJoinResponse.reject();
        }
        customers.put(actor.actorId(), actor);
        System.err.printf(
            "deliverydispatch tracking spot: joined delivery=%s customer=%s%n",
            join.deliveryId(),
            actor.actorId());
        return ZLinkSpotActorJoinResponse.accept(
            encode(new Messages.DeliverySpotJoined(join.deliveryId(), actor.actorId())));
    }

    @Override
    public void onLeaveActor(ZLinkActor actor, CancellationToken cancellationToken) {
        customers.remove(actor.actorId());
    }

    public void record(Messages.DeliveryStatusChanged status) {
        history.add(status);
    }

    private <T> T decode(Message request, Class<T> type) {
        try {
            return json.readValue(request.toByteArray(), type);
        } catch (IOException ex) {
            throw new IllegalArgumentException("Failed to decode " + type.getSimpleName() + ".", ex);
        }
    }

    private Message encode(Object reply) {
        try {
            return Message.from(json.writeValueAsBytes(reply));
        } catch (JsonProcessingException ex) {
            throw new IllegalArgumentException("Failed to encode delivery spot reply.", ex);
        }
    }
}

File: framework/languages/java/samples/kotlin/DeliveryDispatch/Server/Tracking/src/main/kotlin/systems/zlink/samples/kotlin/deliverydispatch/server/tracking/spots/DeliveryTrackingSpot.kt
package systems.zlink.samples.kotlin.deliverydispatch.server.tracking.spots

import com.fasterxml.jackson.databind.ObjectMapper
import systems.zlink.framework.CancellationToken
import systems.zlink.framework.kotlin.ZLinkSuspendingSpot
import systems.zlink.framework.messaging.ZLinkMessage
import systems.zlink.framework.spots.ZLinkSpotActorJoinResponse
import systems.zlink.framework.spots.ZLinkSpotContext
import systems.zlink.framework.spots.ZLinkSpotCreateResponse
import systems.zlink.samples.kotlin.deliverydispatch.server.tracking.actors.CustomerActor
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.DeliverySpotCreate
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.DeliverySpotCreated
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.DeliverySpotJoin
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.DeliverySpotJoined
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.DeliveryStatusChanged

class DeliveryTrackingSpot(
    private val context: ZLinkSpotContext,
    private val directory: DeliverySpotDirectory,
    private val json: ObjectMapper,) : ZLinkSuspendingSpot<CustomerActor>() {
    private val customers = LinkedHashMap<String, CustomerActor>()
    private val history = mutableListOf<DeliveryStatusChanged>()
    private var deliveryId: String = ""

    override fun context(): ZLinkSpotContext = context

    override suspend fun onCreateSuspending(request: ZLinkMessage): ZLinkSpotCreateResponse {
        val create = request.decode(DeliverySpotCreate::class.java)
        deliveryId = create.deliveryId
        directory.add(deliveryId, this)
        return ZLinkSpotCreateResponse.accept(DeliverySpotCreated(deliveryId))
    }

    override suspend fun onActorJoinSuspending(
        actor: CustomerActor,
        request: ZLinkMessage,
        cancellationToken: CancellationToken,
    ): ZLinkSpotActorJoinResponse {
        val join = request.decode(DeliverySpotJoin::class.java)
        if (join.deliveryId != deliveryId) {
            return ZLinkSpotActorJoinResponse.reject()
        }
        customers[actor.actorId()] = actor
        System.err.println(
            "deliverydispatch tracking spot: joined delivery=${join.deliveryId} customer=${actor.actorId()}",
        )
        return ZLinkSpotActorJoinResponse.accept(DeliverySpotJoined(join.deliveryId, actor.actorId()))
    }

    override fun onLeaveActor(actor: CustomerActor, cancellationToken: CancellationToken) {
        customers.remove(actor.actorId())
    }

    fun record(status: DeliveryStatusChanged) {
        history.add(status)
    }
}
